### PR TITLE
TYPING : Series.name -> Optional[Hashable]

### PR DESCRIPTION
**Pull request by simonjayhawkins**
xref https://github.com/pandas-dev/pandas/pull/29062#discussion_r337644656

cc @WillAyd @TomAugspurger

### CODE REVIEW

**Review by TomAugspurger — APPROVED** [review]

**Comment by WillAyd** [comment]
No problem with changing but what caused this issue exactly? Something we need to account for with pickle compat?

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -926,7 +927,7 @@ def aggregate(self, func=None, *args, **kwargs):
 
     agg = aggregate
 
-    def _iterate_slices(self) -> Iterable[Tuple[Hashable, Series]]:
+    def _iterate_slices(self) -> Iterable[Tuple[Optional[Hashable], Series]]:
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: note there is a similar definition for SeriesGroupBy - not sure if that needs to change as well

**Comment by simonjayhawkins** [comment]
> No problem with changing but what caused this issue exactly? Something we need to account for with pickle compat?

no. with check_untyped_defs (i.e. checks body of _unpickle_series_compat)..

`None` is assigned to `name`: `index, name = own_state[0], None`

and then assigned to `self.name`: `self.name = name`

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py]: maybe IndexLabel = Optional[Hashable] in _typing.py

**Review by WillAyd — APPROVED** [review]
lgtm

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**